### PR TITLE
Use Vector2 for position and size on PositionComponent

### PR DESCRIPTION
**Pull request by spydon**
# Description

Make use of Vector2 for position and size on PositionComponent (and some other places).
Becomes extremely handy when everything is vectors when you want to do vector operations.

## Type of change

Please delete options that are not relevant.

- [x] Breaking change (fix or feature that would cause existing functionality to not work as expected)
- [x] This change requires a documentation update

# Checklist:

If something is unclear, please submit the PR anyways and ask about what you thought was unclear.

- [x] I have added an entry under `[next]` in `CHANGELOG.md`
- [x] I have formatted my code with `flutter format`
- [x] I have made corresponding changes to the documentation
- [x] The continuous integration (CI) is passing

### CODE REVIEW

**Comment by spydon** [comment]
This can be pointed to v1.0.0 once #472 has been merge.

File: doc/components.md
Comment (luanpotter) [inline doc/components.md:16]: OMG all is awesome!!!!!!!!

File: doc/components.md
Comment (luanpotter) [inline doc/components.md:20]: nit: `0.0` not `0,0`
nit: use all :)

File: doc/components.md
Comment (luanpotter) [inline doc/components.md:17]: comment doesn't make sense anymore

File: doc/components.md
Comment (luanpotter) [inline doc/components.md:57]: nit: remove new

File: doc/components.md
Comment (luanpotter) [inline doc/components.md:58]: nit: `all` ❤️

File: doc/components.md
Comment (luanpotter) [inline doc/components.md:65]: all :p

File: doc/components.md
Comment (luanpotter) [inline doc/components.md:70]: why remove the svg component doc??

File: doc/components.md
Comment (luanpotter) [inline doc/components.md:71]: isn't Flare on a separated package now? cc @renancaraujo

File: doc/examples/aseprite/lib/main.dart
Comment (luanpotter) [inline doc/examples/aseprite/lib/main.dart:27]: vector.all :)

File: doc/examples/effects/combined_effects/lib/square.dart
Comment (luanpotter) [inline doc/examples/effects/combined_effects/lib/square.dart:12]: `Vector.all(100)`?

File: doc/examples/gestures/lib/main_tapables.dart
Comment (luanpotter) [inline doc/examples/gestures/lib/main_tapables.dart:25]: nit: x then y :p

File: doc/examples/sound/lib/main.dart
Comment (luanpotter) [inline doc/examples/sound/lib/main.dart:55]: all

File: lib/components/flare_component.dart
Comment (luanpotter) [inline lib/components/flare_component.dart:33]: can't we overide a custom setter to the size property as it was before with (width, height)?

File: lib/components/position_component.dart
Comment (luanpotter) [inline lib/components/position_component.dart:46]: we can have setters for width height as well

File: lib/components/position_component.dart
Comment (luanpotter) [inline lib/components/position_component.dart:90]: we could have an anchorPositon getter that returns `position - anchor.relativePositon.multiply(size)`
and then this method becomes `return anchorPositon.toRect(size)`

Comment (luanpotter) [reply lib/components/position_component.dart:90]: we could even have a setter for anchorPosition that does that right math

File: lib/components/sprite_component.dart
Comment (luanpotter) [inline lib/components/sprite_component.dart:26]: since we have Vector2.all we could even drop this constructor. not sure, just an idea. the other one then could fromImagePath instead of reactangle

File: lib/extensions/rect.dart
Comment (luanpotter) [inline lib/extensions/rect.dart:28]: this comment is **not** what the method does!!!
it's LFT**WH** not LF**RB** :)

also, I thought we had start.toRect(end) for this

File: lib/extensions/vector2.dart
Comment (luanpotter) [inline lib/extensions/vector2.dart:22]: nit: origin

File: lib/extensions/vector2.dart
Comment (luanpotter) [inline lib/extensions/vector2.dart:23]: I think `toPositionedRect` and `toRect` here are not 100% clear. I'm not sure what would be best, maybe `startRectWithSize` and `asSizeRect` ?

File: lib/sprite_animation.dart
Comment (luanpotter) [inline lib/sprite_animation.dart:92]: `texturePosition + Vector(i % amountPerRow, i ~/ amountPerRow).multiply(textureSize)`

File: lib/sprite_animation.dart
Comment (luanpotter) [inline lib/sprite_animation.dart:117]: ditto. also, can this be a method? it's repeated twice

File: lib/sprite_animation.dart
Comment (luanpotter) [inline lib/sprite_animation.dart:145]: why this change? this might be a breaking change, we must see how aseprite stores data

File: lib/spritesheet.dart
Comment (luanpotter) [inline lib/spritesheet.dart:40]: not sure if we can have extension constructors, but would be nice to have a Vector2.fromInts like we had for Position

File: lib/sprite_animation.dart
Comment (erickzanardo) [reply lib/sprite_animation.dart:117]: The DRY problem is solved on my PR about asset loading

**Review by erickzanardo — APPROVED** [review]
LGTM, once ALL the comments from @luanpotter are solved :P

File: doc/components.md
Comment (spydon) [reply doc/components.md:20]: I think this one could stay, just so that people understand that it is x, y, but I'll change most of the rest. :)

File: doc/components.md
Comment (spydon) [reply doc/components.md:70]: Oops, re-added it!

File: doc/components.md
Comment (spydon) [reply doc/components.md:71]: We keep all the docs here anyways right?

File: lib/components/flare_component.dart
Comment (spydon) [reply lib/components/flare_component.dart:33]: Fixed!

File: lib/components/position_component.dart
Comment (spydon) [reply lib/components/position_component.dart:46]: Good idea

File: lib/extensions/rect.dart
Comment (spydon) [reply lib/extensions/rect.dart:28]: Very true, and the other one we already have `size.toRect()` for.
Removing both.

File: lib/sprite_animation.dart
Comment (spydon) [reply lib/sprite_animation.dart:117]: I converted them, but swapped back. I think it is easier to read like it is now compared to this:
```dart
      final position = texturePosition +
          Vector2(i % amountPerRow.toDouble(), (i ~/ amountPerRow).toDouble())
            ..multiply(textureSize);
```

File: lib/spritesheet.dart
Comment (spydon) [reply lib/spritesheet.dart:40]: We can not, unfortunately.

File: lib/sprite_animation.dart
Comment (spydon) [reply lib/sprite_animation.dart:145]: It is just JSON right, and since there only is one number type there it should be fine?

File: lib/extensions/vector2.dart
Comment (spydon) [reply lib/extensions/vector2.dart:23]: I agree that they aren't that clear, I'm not convinced that your suggestions are more clear though.

File: lib/components/position_component.dart
Comment (spydon) [reply lib/components/position_component.dart:90]: Much cleaner, good suggestion.

**Comment by spydon** [comment]
I will update the anchorPosition och position comments tonight.

**Review by renancaraujo — APPROVED** [review]
Looking great after the renaming.

File: lib/components/position_component.dart
Comment (luanpotter) [inline lib/components/position_component.dart:57]: nit: relativePosition creates a fresh copy each time, so I think you don't need to clone here

File: lib/components/position_component.dart
Comment (luanpotter) [inline lib/components/position_component.dart:101]: nice impl!

File: lib/components/position_component.dart
Comment (luanpotter) [inline lib/components/position_component.dart:130]: ditto

File: lib/extensions/vector2.dart
Comment (luanpotter) [reply lib/extensions/vector2.dart:23]: yeah I think we don't have the perfect one yet. I am ok with merging for now but let's keep thinking cc @renancaraujo @erickzanardo @wolfenrain if you have ideas

File: lib/sprite_animation.dart
Comment (luanpotter) [reply lib/sprite_animation.dart:117]: true, I forgot the toDouble. we should make a Vector2.fromInts I think (like we had for Position). but that is a followup for sure.

File: lib/sprite_animation.dart
Comment (luanpotter) [reply lib/sprite_animation.dart:145]: there is only one number type in JSON but I am not sure how dart convert it to map. I am scared dart will throw an error (unlike js in dart int/double are distinct types). I think it might be safer to cast to int and call `toDouble()`, since it's a dynamic map. unless we can easily test it or there is an example and it works.

File: lib/spritesheet.dart
Comment (luanpotter) [reply lib/spritesheet.dart:41]: 😢 

we can do a static function then like Vector2Extension.fromInts. but let's think that through for a followup

**Review by luanpotter — APPROVED** [review]
LGTM with final nits

**Comment by spydon** [comment]
> LGTM with final nits

Updated according to nits.
I'm not convinced by the static fromInts method on Vector2Extension, since it can't be used directly from Vector2.

File: lib/extensions/vector2.dart
Comment (wolfenrain) [reply lib/extensions/vector2.dart:23]: It of course also depends on the motivation behind these methods. I can understand the `toPositionedRect`, tho I would indeed name it differently. But the `toRect` looks "useless" to me without context. 

So what are the usecases, and from there we can think of naming them, fitting with the context.

Comment (spydon) [reply lib/extensions/vector2.dart:23]: `toRect` is used for example to convert the size vector to a `Rect` that the canvas can use.
Any ideas of what to name the `toPositionedRect`, and `toRect`?

**Comment by luanpotter** [comment]
Yeah, I will think of something btter. that is definitely a followup though

**Pull request merged**
